internal/auth/configuration.go

```diff
@@ -0,0 +1,437 @@
+	ServerConfig    ServerConfig              `mapstructure:"server"`
+	MetricsConfig   MetricsConfig             `mapstructrue:"metrics"`
+	LoggingConfig   LoggingConfig             `mapstructure:"logging"`
+}
+
+func (c Configuration) Validate() error {
+	for slug, providerConfig := range c.ProviderConfigs {
+		if err := providerConfig.Validate(); err != nil {
+			return xerrors.Errorf("invalid provider.%s config: %w", slug, err)
+		}
+	}
+
+	for slug, clientConfig := range c.ClientConfigs {
+		if err := clientConfig.Validate(); err != nil {
+			return xerrors.Errorf("invalid client.%s config: %w", slug, err)
+		}
+	}
+
+	if err := c.SessionConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid session config: %w", err)
+	}
+
+	if err := c.ServerConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid server config: %w", err)
+	}
+
+	if err := c.AuthorizeConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid authorize config: %w", err)
+	}
+
+	if err := c.MetricsConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid metrics config: %w", err)
+	}
+
+	return nil
+}
+
+type ProviderConfig struct {
+	ProviderType string       `mapstructure:"type"`
+	ProviderSlug string       `mapstructure:"slug"`
+	ClientConfig ClientConfig `mapstructure:"client"`
+	Scope        string       `mapstructure:"scope"`
+
+	// provider specific
+	GoogleProviderConfig GoogleProviderConfig `mapstructure:"google"`
+	OktaProviderConfig   OktaProviderConfig   `mapstructure:"okta"`
+
+	// caching
+	GroupCacheConfig GroupCacheConfig `mapstructure:"groupcache"`
+}
+
+func (pc ProviderConfig) Validate() error {
+	if pc.ProviderType == "" {
+		return xerrors.Errorf("invalid provider.type: %q", pc.ProviderType)
+	}
+
+	// TODO: more validation of provider slug, should conform to simple character space
+	if pc.ProviderSlug == "" {
+		return xerrors.Errorf("invalid provider.slug: %q", pc.ProviderSlug)
+	}
+
+	if err := pc.ClientConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid provider.client: %w", err)
+	}
+
+	switch pc.ProviderType {
+	case "google":
+		if err := pc.GoogleProviderConfig.Validate(); err != nil {
+			return xerrors.Errorf("invalid provider.google config: %w", err)
+		}
+	case "okta":
+		if err := pc.OktaProviderConfig.Validate(); err != nil {
+			return xerrors.Errorf("invalid provider.okta config: %w", err)
+		}
+	case "test":
+		break
+	default:
+		return xerrors.Errorf("unknown provider.type: %q", pc.ProviderType)
+	}
+
+	if err := pc.GroupCacheConfig.Validate(); err != nil {
+		return xerrors.Errorf("invalid provider.groupcache config: %w", err)
+	}
+
+	return nil
+}
+
+type GoogleProviderConfig struct {
+	Credentials    string `mapstructure:"credentials"`
+	Impersonate    string `mapstructure:"impersonate"`
+	ApprovalPrompt string `mapstructure:"prompt"`
+}
+
+func (gpc GoogleProviderConfig) Validate() error {
+	// TOOD: what can we validate here?
+	return nil
```

Comment:
For the Okta provider, it seems like we actually test for the `OrgURL` and `ServerID` both here _and_ in the actual provider itself. 

In a similar fashion, the Google provider tests for `adminEmail` and `credsFilePath` - I'm not sure if we want to have it just be tested in once place, or in both places? (But, I guess we should be consistent with both providers with whichever direction we go)

Reply:
For Okta, those variable are mandatory. For Google, they are optional.